### ISSUE TITLE
Resource Managment for Hyperparameter Tuning with Ray Tune??

### ISSUE THREAD

**Issue by HappyMistery** [ultralytics/ultralytics#20234]
I am using the following code to do some hyperparameter tuning with YOLO11 and Ray Tune:
```
search_space = {
    "lr0": tune.loguniform(1e-5, 1e-1),
    "momentum": tune.uniform(0.6, 0.98),
    "weight_decay": tune.uniform(0.0, 0.001),
}

tune_config = tune.TuneConfig(   # Not really sure if this has to be specified
    num_samples=10,
    max_concurrent_trials=10
)

result_grid = model.tune(
    data=os.path.abspath("data.yaml"),
    space=search_space,
    use_ray=True,
    gpu_per_trial = 0,
    optimizer = "SGD",
    device=0,
    epochs=100,
    imgsz=416,
    batch=1,
    workers=6,
    resume=False
)
```

And what I have noticed is that in my case, if I use ``` gpu_per_trial = 0, ``` Ray tune uses 8/12 CPUs, 0/1 GPUs, but if I use ``` gpu_per_trial = 1, ``` Ray tune uses 0/12 CPUs, 1/1 GPUs (which obviously makes sense).

Now, I am running for 100 epochs at around 6mins/epoch (both for CPU only usage and GPU only usage) and for 10 trials, so basically my hyperparameter tuning would take around  10h per trial so 100h (4,16 days aprox).

Knowing that in my case I can run a trial with either  8/12 CPUs & 0/1 GPUs or  0/12 CPUs & 1/1 GPUs, I want to know is if there is any way I can use 8CPUs for a trial while at the same time use 1GPU for another trial in order to be able to run 2 trials in parallel and reduce the total time to 50h instead of 100h...
Does anyone know a way to do that?

Thanks :D

# Aditional info
Ultralytics 8.3.111 🚀 Python-3.11.9 torch-2.6.0+cu118 CUDA:0 (NVIDIA GeForce RTX 3060, 12288MiB)
Setup complete ✅ (12 CPUs, 15.9 GB RAM, 410.7/476.1 GB disk)

OS                  Windows-10-10.0.26100-SP0
Environment         Windows
Python              3.11.9
Install             git
Path                C:\UNI\URV\TFG\.venv\Lib\site-packages\ultralytics
RAM                 15.89 GB
Disk                410.7/476.1 GB
CPU                 AMD Ryzen 5 5600X 6-Core Processor
CPU count           12
GPU                 NVIDIA GeForce RTX 3060, 12288MiB
GPU count           1
CUDA                11.8

numpy               ✅ 1.26.4<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.1>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.0.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.15.2>=1.4.1
torch               ✅ 2.6.0+cu118>=1.8.0
torch               ✅ 2.6.0+cu118!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.21.0+cu118>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 7.0.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0

**Comment by UltralyticsAssistant** [comment]
👋 Hello @HappyMistery, thank you for your detailed question and for using Ultralytics 🚀! This is an automated response to help get you started—an Ultralytics engineer will also assist you soon.

We recommend visiting the [Docs](https://docs.ultralytics.com/) for helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples that may address similar questions.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If your question relates to custom training or hyperparameter tuning, please include as much detail as possible, such as training logs, dataset examples, and confirm you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community on your favorite platform! For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧. For deeper discussions, check out [Discourse](https://community.ultralytics.com/), or join our [Subreddit](https://reddit.com/r/Ultralytics) to share insights with other users.

## Upgrade

Please ensure you are using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date, verified environments (with all dependencies such as [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience while we review your request! If you have any additional details or a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/), please add them to help us assist you faster. 🛠️

**Comment by Y-T-G** [comment]
It's a question related to Ray Tune, so it's better asked on Ray Tune GitHub repo.

But from their docs, this doesn't seem possible.

https://docs.ray.io/en/latest/tune/tutorials/tune-resources.html

**Comment by HappyMistery** [comment]
Thanks, will try there :)

**Comment by glenn-jocher** [comment]
You're welcome, @HappyMistery! Just to add a bit more context, while our basic Ray Tune integration lets you control `gpu_per_trial`, the advanced resource partitioning you're looking for would indeed require Ray-specific configuration. 

For your specific use case, you might want to explore Ray's custom resource handling with `tune.with_resources()` which can give you more granular control than our wrapper provides. The Ray Tune team should be able to help with the exact implementation for your mixed CPU/GPU parallel trials approach.

If you find a good solution, we'd appreciate if you could share it back here as it might benefit other users with similar requirements!
